### PR TITLE
Add working read alignment output for MALT/Kraken2/Centrifuge/DIAMOND

### PR DESCRIPTION
**Pull request by jfy133**
Adds support for saving aligned reads if the tool supports this

To close #76 

### New Functionality

- Updates modules to allow saving of alignments (and changed output channel names accordingly)
- Adds support for saving alignments for MALT/Kraken2/Centrifuge alongisde taxonomic profiling
- Adds support for DIAMOND to save alignmetns (but this cannot also save OTU table as well, so provide warning only alignments will be saved when giving flag)

###  Additional

- Limits to MEGAN_RMA2INFO to one job at a time in the test profile (uses same code base as MALT, for which we already implemented the maxFork because of locking up of machines, and was causing problems for me in tests)
- Fixes a typo for `test_nopreprocessing`
- Fixes typo in `malt_generate_megansummary`
- Removes a left over `.dump(tag: filtlong)` (picked up when I was cleaning up debugging stuff for alignment-saving
- Fixes MALT processes not having a `id` tag causing file publication in `null`

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/taxprofiler/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/taxprofiler _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: subworkflows/local/profiling.nf [hunk 1]
```diff
@@ -65,12 +65,18 @@ workflow PROFILING {
...
+                                        // TODO wut? [9a/a441d6] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:MALT_RUN (null)
+                                        def temp_meta = [ id: meta['db_name'] ]
+                                        def new_db_meta = db_meta.clone()
+                                        new_db_meta['db_params'] = db_meta['db_params'] + sam_format
+                                        def new_meta = temp_meta + new_db_meta
+                                        new_meta['id'] = new_meta['db_name']
+                                        [ new_meta, reads, db ]
                                 }
                                 .groupTuple(by: [0,2])
+                                .dump(tag: "into_malt")
```
Comment (jfy133) [inline subworkflows/local/profiling.nf:79]:
```suggestion
```

File: subworkflows/local/profiling.nf [hunk 2]
```diff
@@ -65,12 +65,18 @@ workflow PROFILING {
         ch_input_for_malt =  ch_input_for_profiling.malt
                                 .filter { it[0]['instrument_platform'] == 'ILLUMINA' }
                                 .map {
-                                    it ->
-                                        def temp_meta =  [ id: it[2]['db_name']]  + it[2]
-                                        def db = it[3]
-                                        [ temp_meta, it[1], db ]
+                                    meta, reads, db_meta, db ->
+                                        def sam_format = params.malt_save_reads ? ' --alignments ./ -za false' : ""
+                                        // TODO wut? [9a/a441d6] Submitted process > NFCORE_TAXPROFILER:TAXPROFILER:PROFILING:MALT_RUN (null)
```
Comment (jfy133) [inline subworkflows/local/profiling.nf:70]:
```suggestion
```

File: conf/modules.config [hunk 3]
```diff
@@ -289,7 +289,7 @@ process {
         publishDir = [
             path: { "${params.outdir}/kraken2/${meta.db_name}" },
             mode: params.publish_dir_mode,
-            pattern: '*.{txt}'
+            pattern: '*.{txt,report,fastq.gz}'
```
Comment (sofstam) [inline conf/modules.config:292]: The pattern in the kraken module is:

  ```
   type: file
      description: |
        Kraken2 report containing stats about classified
        and not classifed reads.
      pattern: "*.{report.txt}"
```

Should this be:

`pattern: '*.{txt,fastq.gz}'`

File: nextflow.config [hunk 4]
```diff
@@ -102,16 +102,17 @@ params {
...
     malt_mode                  = 'BlastN'
-    malt_generatemegansummary = false
+    malt_generate_megansummary = false
+    malt_save_reads            = false
 
     // kraken2
-    run_kraken2                = false
+    run_kraken2                     = false
+    kraken2_save_reads              = false
+    kraken2_save_readclassification = false
```
Comment (sofstam) [inline nextflow.config:111]: Maybe I misunderstand something here, but what is the difference between `kraken2_save_reads` and `kraken2_save_readclassification`?

File: subworkflows/local/profiling.nf [hunk 5]
```diff
@@ -65,10 +65,14 @@ workflow PROFILING {
         ch_input_for_malt =  ch_input_for_profiling.malt
                                 .filter { it[0]['instrument_platform'] == 'ILLUMINA' }
                                 .map {
-                                    it ->
-                                        def temp_meta =  [ id: it[2]['db_name']]  + it[2]
-                                        def db = it[3]
-                                        [ temp_meta, it[1], db ]
+                                    meta, reads, db_meta, db ->
```
Comment (sofstam) [inline subworkflows/local/profiling.nf:68]: Not sure I understand exactly this chunk of code but maybe someone else might review this as well?

File: conf/modules.config (hunk 3 above)
Comment (jfy133) [reply conf/modules.config:292]: txt is the default output, report is the `readclassification` thing (basically gives a list of all reads, and what taxonomic assignment each read was provided), and fastqc.gz is the aligned reads

File: subworkflows/local/profiling.nf (hunk 5 above)
Comment (jfy133) [reply subworkflows/local/profiling.nf:68]: I'll add somecomments :)

Comment (sofstam) [reply subworkflows/local/profiling.nf:68]: Thanks for the comments.

File: conf/modules.config (hunk 3 above)
Comment (jfy133) [reply conf/modules.config:292]: I don' follow the qeustion, sorry - can you rephrase?

Comment (sofstam) [reply conf/modules.config:292]: I was wondering if the parameter `kraken2_save_readclassification` could be named as `kraken2_save_classified_reads` instead?

Comment (jfy133) [reply conf/modules.config:292]: I don't think so, I fear people would think this is actual reads that were classified in e.g. FASTQ format (which is what `save_aligned` is meant to be doing), whereas this is a text file with a list of reads and their taxonomic assingments (and more info about them)

Comment (sofstam) [reply conf/modules.config:292]: Sounds good then.

Comment (jfy133) [reply conf/modules.config:292]: AAwesome, I'll re-request review for you to check you're happy with it now I've added the comments :)

**Review by sofstam — APPROVED** [review]
Looks good to me, good job :)
